code/modules/mob/identity.dm

```diff
@@ -0,0 +1,578 @@
+
+/datum/mind/proc/open_idman()
+	if(!current)
+		return
+	if(!idman)
+		idman = new(src)
+	idman.ui_interact(current)
+
+/mob/living/carbon/human/default_identity_heard()
+	. = "Unknown"
+
+/mob/living/carbon/human/default_identity_seen(default_to_heard=TRUE)
+	if(w_uniform && !(wear_suit && (wear_suit.flags_inv & HIDEJUMPSUIT)) && w_uniform.identity_name)
+		. = w_uniform.identity_name
+	else if(wear_suit && wear_suit.identity_name)
+		. = wear_suit.identity_name
+	else if(default_to_heard)
+		. = default_identity_heard()
+
+/mob/living/carbon/human/default_identity_interact()
+	. = get_id_name("")
+	if(!.)
+		var/hair_text
+		var/seen_text = default_identity_seen(FALSE)
+		var/accessory_text
+		// eyecolor, haircolor+hairstyle, hat or mask, jobname
+		if(!hair_covered() && hair_color && hair_style)
+			var/htext_color = consonants(color_hex2color(hair_color))
+			var/htext_style = consonants(hair_style)
+			var/datum/sprite_accessory/hair/tortoise = hair_styles_list[hair_style]
+			if(htext_color && tortoise && tortoise.icon_state)
+				hair_text = "[htext_color] [htext_style]"
+			else
+				hair_text = "[htext_style]"
+		if(head && head.name)
+			accessory_text = " [consonants(head.name)]"
+		else if(wear_mask && wear_mask.name)
+			accessory_text = " [consonants(wear_mask.name)]"
+		. = "[hair_text][seen_text ? " [seen_text]" : ""][accessory_text]"
+
+/mob/living/carbon/human/get_faceprint()
+	if(dna)
+		. = dna.uni_identity
+
+/datum/identity_manager
+	var/cat = CATEGORY_VOICEPRINTS
+	var/datum/mind/mind
+	var/list/print_refs = list()
+	var/list/refs_lookup = list()
+	var/selected_ref
+	var/select_mode
+
+/datum/identity_manager/New(datum/mind/the_mind)
+	if(the_mind && istype(the_mind))
+		mind = the_mind
+	else
+		qdel(src)
+
+/datum/identity_manager/proc/done_selecting()
+	if(select_mode == IDMAN_MODE_LINK)
+		cat = mind.inverse_category(cat)
+	selected_ref = null
+	select_mode = null
+
+/datum/identity_manager/proc/unlink(print)
+	var/list/print_entry = mind.get_print_entry(print, cat)
+	if(!print_entry)
+		return
+	var/linked_cat = mind.inverse_category(cat)
+	var/linked_print = print_entry[4]
+	if(linked_print)
+		var/list/linked_print_entry = mind.get_print_entry(linked_print, linked_cat)
+		if(linked_print_entry && linked_print_entry[4] == print)
+			linked_print_entry[4] = null
+			mind.set_print_entry(linked_print, linked_print_entry, linked_cat)
+	print_entry[4] = null
+	mind.set_print_entry(print, print_entry, cat)
+
+/datum/identity_manager/proc/delete_print(ref)
+	var/print = print_refs[ref]
+	refs_lookup -= print
+	print_refs -= ref
+	var/list/print_entry = mind.get_print_entry(print, cat)
+	if(!print_entry)
+		return
+	unlink(print)
+	mind.set_print_entry(print, null, cat)
+	if(selected_ref == ref)
+		done_selecting()
+
+/datum/identity_manager/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/ui_state/state = alive_state)
+	if(!mind)
+		qdel(src)
+		return
```

Comment:
Again, unless you can guarantee Destroy will clean up references, this is probably not safe